### ISSUE TITLE
Using Json.NET library suggestion

### ISSUE THREAD

**Issue by hzahed** [thepirat000/Audit.NET#74]
First of all, thank you for the fantastic library! 
I have a suggestion for the creation of Bson Document in the following events in Audit.Net MongoDB provider:

- InsertEvent
- ReplaceEvent
- Serialize<T>

We are working on a WebApi Core 2.0 project with MSSQL database and we use MongoDB to save our application logs.

The point is that for saving complex entities with navigation properties we face the following error while saving the records to MongoDB:
BsonSerializationException: Maximum serialization depth exceeded (does the object being serialized have a circular reference?)
This happened because the ToBsonDocument() method can't Ignore Reference Loopings and always return an exception.

We had a quick fix for this issue as followed.
Changing this part in InsertEvent/ReplaceEnevt:
`var doc = auditEvent.ToBsonDocument();`
to this:
`var doc = BsonDocument.Parse(JsonConvert.SerializeObject(auditEvent, new JsonSerializerSettings
 {
     ReferenceLoopHandling = ReferenceLoopHandling.Ignore
 }));`

and also this part in Serialize<T>:
`return value.ToBsonDocument(typeof(object));`
to this:
`return BsonDocument.Parse(JsonConvert.SerializeObject(value, new JsonSerializerSettings
 {
     ReferenceLoopHandling = ReferenceLoopHandling.Ignore
 }));`

OR

You may add a setting to the UseMongoDB configuration so that Ignoring Reference Looping can be enabled or disabled.

Any suggestion or additional configuration would be appreciated.

**Comment by thepirat000** [comment]
I think the cleanest way to implement this would be to provide a way to inject the BsonParser (a function that takes an AuditEvent
and returns a BsonDocument) and the Serializer (a function that takes any object and serializes it to another object)
The last one is needed for the pre-serialization of custom fields and target object.

So for your case, you could setup like this:

```c#
var defaultSettings = new JsonSerializerSettings { ReferenceLoopHandling = ReferenceLoopHandling.Ignore };
Audit.Core.Configuration.Setup()
    .UseMongoDB(config => config
        .ConnectionString("mongodb://localhost:27017")
        .Database("Audit")
        .Collection("Event")
        .CustomParser(o => 
            BsonDocument.Parse(JsonConvert.SerializeObject(o, defaultSettings))
        )
        .CustomSerializer(o =>
            JsonConvert.DeserializeObject(JsonConvert.SerializeObject(o, defaultSettings))
        ))
```

I will work on this change and let you know

**Comment by hzahed** [comment]
Both options seems to be fine fore me but injecting a custom provider offers more flexibility!

**Comment by thepirat000** [comment]
I'll better go with your serialization method by default, but will allow the user to change the `JsonSerializerSettings`, i.e.:

```c#
Audit.Core.Configuration.Setup()
    .UseMongoDB(config => config
        .ConnectionString("mongodb://localhost:27017")
        .Database("Audit")
        .Collection("Event")
        .CustomSerializerSettings(new JsonSerializerSettings() {  TypeNameHandling = TypeNameHandling.All })));
```

**Comment by hzahed** [comment]
That would be great!

**Comment by thepirat000** [comment]
This is included from version 9.0.0.

Feel free to re-open the issue is you're still having problems.
